### PR TITLE
Debug instance for the gossipsub behaviour

### PR DESCRIPTION
**Pull request by rklaehn**
To assist looking into https://github.com/libp2p/rust-libp2p/issues/1671

Not sure if the nicer debug output is worth the extra dependency. For me it is, but YMMV. I can remove this if you see this differently.

Edit: I added truncation of the message bytes to ensure that the debug output for a gossipsub msg is not too big.

### CODE REVIEW

File: protocols/gossipsub/src/config.rs [hunk 1]
```diff
@@ -231,8 +231,19 @@ impl GossipsubConfigBuilder {
...
+                if let Ok(text) = std::str::from_utf8(self.0) {
+                    write!(f, "\"{}\"", text)
+                } else {
+                    write!(f, "{}", hex_fmt::HexFmt(self.0))
+                }
+            }
+        }
         let mut builder = f.debug_struct("GossipsubConfig");
-        let _ = builder.field("protocol_id", &self.protocol_id);
+        let _ = builder.field("protocol_id", &StringOrBytes(&self.protocol_id));
```
Comment (tomaka) [inline protocols/gossipsub/src/config.rs:246]: It feels more simple to me to just inline that code.

(I typed that blindly, might not compile)

```suggestion
        if let Ok(text) = std::str::from_utf8(self.protocol_id.0) {
            let _ = builder.field("protocol_id", &text));
        } else {
            let _ = builder.field("protocol_id", hex_fmt::HexFmt(self.protocol_id.0));
        }
```

File: protocols/gossipsub/src/config.rs [hunk 2]
```diff
@@ -231,8 +231,19 @@ impl GossipsubConfigBuilder {
...
+
+        impl<'a> std::fmt::Debug for StringOrBytes<'a> {
+            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                if let Ok(text) = std::str::from_utf8(self.0) {
+                    write!(f, "\"{}\"", text)
+                } else {
+                    write!(f, "{}", hex_fmt::HexFmt(self.0))
+                }
+            }
+        }
```
Comment (tomaka) [inline protocols/gossipsub/src/config.rs:244]:
```suggestion
```

File: protocols/gossipsub/src/config.rs (hunk 1 above)
Comment (rklaehn) [reply protocols/gossipsub/src/config.rs]: You are right. I had that thing laying around and copied it in, but here it is only used once.

**Comment by romanb** [comment]
@rklaehn If you could update the gossipsub changelog and merge master, I can go ahead and merge this.

**Pull request merged**